chain/chain/src/chain.rs

```diff
@@ -781,8 +781,15 @@ impl Chain {
         } else {
             debug!(target: "chain", "Downloading state for {:?}, I'm {:?}", shards_to_state_sync, me);
 
+            // This is not the way protocol features are normally gated, but this is not actually
+            // a protocol feature/change. The protocol version is just an easy way to coordinate
+            // among nodes for now
+            let protocol_version =
+                self.epoch_manager.get_epoch_protocol_version(block.header().epoch_id())?;
+            let sync_hash =
+                if protocol_version < 72 { *block.header().hash() } else { CryptoHash::default() };
```

Comment:
I agree that magic values are ugly in general, but here it’s done to avoid the need for a database migration. 

So before this change, this field is always set to the hash of the first block of the epoch (which is the sync_hash we use for state sync). And this is also used as the key we look up this value with. But this `sync_hash` field in the `StateSyncInfo` struct is actually only ever checked once, where we just [check if it’s equal to the key that looked up that struct](https://github.com/near/nearcore/blob/70acea66bac9e5eeee64c0f417112f762e2a2b55/chain/client/src/client.rs#L2467).

So this `sync_hash` field is actually just redundant data and is not used meaningfully, which is why I took the opportunity to change the way we set it, because it can be done with logic that’s consistent before/after the binary upgrade. We just set it to zero after the protocol change, and if we read it and that field is set, then that’s the sync hash we want to use, otherwise, we’re still looking for the sync hash.

So all this is just to avoid a DB migration. I felt that if we could do so, it’s not so bad to accept slightly ugly code with a magic value of zeros, because a DB migration is a sort of nonzero development/UX churn point. But wdyt? I could of course just do this the right way and add a DB migration if we don’t think it’s such a big deal

Reply:
Thanks for sharing context, I didn't realize it would necessitate a db migration. 

Still I think state sync being in a pretty bad place as it is I wouldn't want to add more tech debt to it. I don't think a db migration is a big deal even if it takes a bit extra effort. I would expect the state sync info to be minimal in size so the migration should be really quick too. Can you try and see how bad is it?